core/unix/signal.c

```diff
@@ -3235,20 +3235,19 @@ thread_set_self_context(void *cxt, bool is_detach_external)
     ASSERT_NOT_IMPLEMENTED(false); /* PR 405694: can't use regular sigreturn! */
 #endif
     memset(&frame, 0, sizeof(frame));
-#if defined(X86)
-    dcontext_t *dcontext = get_thread_private_dcontext();
-#endif
 #ifdef LINUX
+    frame.uc.uc_mcontext = *sc;
 #    ifdef X86
-    byte *xstate = get_and_initialize_xstate_buffer(dcontext);
-    frame.uc.uc_mcontext.fpstate = &((kernel_xstate_t *)xstate)->fpstate;
+    dcontext_t *dcontext = get_thread_private_dcontext();
+    frame.uc.uc_mcontext.fpstate = (kernel_fpstate_t *)get_and_initialize_xstate_buffer(
```

Comment:
Based on the other usage of get_and_initialize_xstate_buffer, I'm pretty sure it's fine to set .fpstate straight to this, and not do the `&((kernel_xstate_t *)xstate)->fpstate;` nonsense. (it appears the fpstate is the first member anyway and kernel_xstate_t and kernel_fpstate_t are treated interchangeably in several places... idk the backstory here)

There is still a duplicate `dr_xgetbv` call... `get_and_initialize_xstate_buffer` and `save_xmm` both do it, but they fill in different fields (`fpstate->sw_reserved.xstate_bv` vs. `xstate_hdr.xstate_bv`).

Reply:
Hmm, I do see the other case that treats it the same way in your new code.  It seems to be the same definition for all arches.  OK.